x-pack/plugin/ml/src/main/java/org/elasticsearch/xpack/ml/datafeed/DatafeedJob.java

```diff
@@ -173,23 +182,79 @@ private void checkForMissingDataIfNecessary() {
             this.lastDataCheckTimeMs = this.currentTimeSupplier.get();
             List<BucketWithMissingData> missingDataBuckets = delayedDataDetector.detectMissingData(latestFinalBucketEndTimeMs);
             if (missingDataBuckets.isEmpty() == false) {
-
                 long totalRecordsMissing = missingDataBuckets.stream()
                     .mapToLong(BucketWithMissingData::getMissingDocumentCount)
                     .sum();
-                // The response is sorted by asc timestamp, so the last entry is the last bucket
-                Date lastBucketDate = missingDataBuckets.get(missingDataBuckets.size() - 1).getBucket().getTimestamp();
-                int newAudit = Objects.hash(totalRecordsMissing, lastBucketDate);
-                if (newAudit != lastDataCheckAudit) {
-                    auditor.warning(jobId,
-                        Messages.getMessage(Messages.JOB_AUDIT_DATAFEED_MISSING_DATA, totalRecordsMissing,
-                            XContentElasticsearchExtension.DEFAULT_DATE_PRINTER.print(lastBucketDate.getTime())));
-                    lastDataCheckAudit = newAudit;
+                Annotation annotation = createAnnotation(missingDataBuckets.get(0).getBucket().getTimestamp(),
+                    missingDataBuckets.get(missingDataBuckets.size() - 1).getBucket().getTimestamp(),
+                    totalRecordsMissing);
+
+                // Have we an annotation that covers the same area with the same message?
+                // Cannot use annotation.equals(other) as that checks createTime
+                if (lastDataCheckAnnotation != null
+                    && annotation.getAnnotation().equals(lastDataCheckAnnotation.getAnnotation())
+                    && annotation.getTimestamp().equals(lastDataCheckAnnotation.getTimestamp())
+                    && annotation.getEndTimestamp().equals(lastDataCheckAnnotation.getTimestamp())) {
+                    return;
+                }
+
+                if (lastDataCheckAnnotationId != null) {
+                    updateAnnotation(annotation);
+                } else {
+                    lastDataCheckAnnotationId = addAndSetDelayedDataAnnotation(annotation);
                 }
             }
         }
     }
 
+    private Annotation createAnnotation(Date startTime, Date endTime, long recordsMissing) {
+       String msg = Messages.getMessage(Messages.JOB_AUDIT_DATAFEED_MISSING_DATA, recordsMissing,
+            XContentElasticsearchExtension.DEFAULT_DATE_PRINTER.print(endTime.getTime()));
+       return new Annotation(msg,
+           new Date(currentTimeSupplier.get()),
+           "ml-delayed-data-checker",
```

Comment:
Maybe `SystemUser.NAME`? I worry that in the future when we do object level security using a username that doesn’t correspond to a real user may be a security hole.

Reply:
🔐 💥